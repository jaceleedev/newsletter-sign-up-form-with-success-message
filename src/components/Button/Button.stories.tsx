import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';
import Button from './Button';

const meta = {
  component: Button,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    type: {
      description: 'Button type (button, submit, reset)',
      control: 'select',
      options: ['button', 'submit', 'reset'],
    },
    label: {
      description: 'Text to be displayed on the button',
      control: 'text',
    },
  },
  args: {
    type: 'submit',
    onClick: fn(),
  },
  decorators: [
    (Story) => (
      <div style={{ width: '327px' }}>
        <Story />
      </div>
    ),
  ],
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: { label: 'Subscribe to monthly newsletter' },
};

export const Hovered: Story = {
  args: { label: 'Subscribe to monthly newsletter' },
  parameters: { pseudo: { hover: true } },
};

export const Pending: Story = {
  args: { label: 'Processing...', disabled: true },
};
