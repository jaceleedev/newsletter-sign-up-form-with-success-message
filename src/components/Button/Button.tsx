'use client';

import { ButtonHTMLAttributes } from 'react';
import { useFormStatus } from 'react-dom';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  type: 'button' | 'submit' | 'reset';
  label: string;
}

function Button({ type, label, ...props }: Readonly<ButtonProps>) {
  const { pending } = useFormStatus();

  return (
    <button
      type={type}
      className="text-body-bold text-white text-center w-full pt-[18px] pb-[14px] rounded-lg bg-dark-navy hover:bg-gradient-primary hover:shadow-primary transition-all"
      disabled={pending}
      aria-busy={pending}
      {...props}
    >
      {pending ? 'Processing...' : label}
    </button>
  );
}

export default Button;
