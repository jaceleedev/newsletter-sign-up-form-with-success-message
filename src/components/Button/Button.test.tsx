import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import Button from './Button';
import { FormStatus, useFormStatus } from 'react-dom';

expect.extend(toHaveNoViolations);

vi.mock('react-dom', async () => {
  return {
    useFormStatus: vi.fn().mockReturnValue({ pending: false }),
  };
});

describe('Button', () => {
  // 1. Rendering test
  it('renders correctly', () => {
    render(<Button type="submit" label="Render test" />);
    const button = screen.getByRole('button', { name: 'Render test' });
    expect(button).toBeDefined();
  });

  // 2. Props test
  it('applies correct type attribute', () => {
    render(<Button type="submit" label="Props test" />);
    const button = screen.getByRole('button', { name: 'Props test' });
    expect(button).toHaveProperty('type');
  });

  // 3. Styles test
  it('applies the correct styles', () => {
    render(<Button type="submit" label="Style test" />);
    const button = screen.getByRole('button', { name: 'Style test' });
    expect(button.className).toContain('text-body-bold');
    expect(button.className).toContain('text-white');
    expect(button.className).toContain('bg-dark-navy');
    expect(button.className).toContain('rounded-lg');
  });

  // 4. State change test
  it('changes label when pending', () => {
    vi.mocked(useFormStatus).mockReturnValueOnce({
      pending: true,
    } as FormStatus);
    render(<Button type="button" label="State change test" />);
    const button = screen.getByRole('button', { name: 'Processing...' });
    expect(button).toBeDefined();
  });

  // 5. Behavior test
  it('calls onClick when clicked', () => {
    const handleClick = vi.fn();
    render(
      <Button type="submit" label="Behavior test" onClick={handleClick} />
    );
    const button = screen.getByRole('button', { name: 'Behavior test' });
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledOnce();
  });

  // 6. Accessibility test
  it('has no accessibility violations', async () => {
    const { container } = render(
      <Button type="button" label="Accessibility test" />
    );
    const result = await axe(container);
    expect(result).toHaveNoViolations();
  });

  // 7. Snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Button type="submit" label="Snapshot test" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
