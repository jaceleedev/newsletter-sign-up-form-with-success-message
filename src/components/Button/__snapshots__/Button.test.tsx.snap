// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-busy="false"
    class="text-body-bold text-white text-center w-full pt-[18px] pb-[14px] rounded-lg bg-dark-navy hover:bg-gradient-primary hover:shadow-primary transition-all"
    type="submit"
  >
    Snapshot test
  </button>
</DocumentFragment>
`;
